### PR TITLE
added HtmlHelper::__call() method

### PR DESCRIPTION
**Pull request by mirko-pagliai**
See https://github.com/cakephp/cakephp/issues/8631

### CODE REVIEW

File: tests/TestCase/View/Helper/HtmlHelperTest.php [hunk 1]
```diff
@@ -1852,6 +1852,32 @@ public function testTableCells()
...
+        $result = $this->Html->h3();
+        $expected = $this->Html->tag('h3');
+        $this->assertEquals($expected, $result);
+        
+        $result = $this->Html->h3('text');
+        $expected = $this->Html->tag('h3', 'text');
+        $this->assertEquals($expected, $result);
+        
+        $result = $this->Html->h3('<text>', ['class' => 'class-name']);
+        $expected = $this->Html->tag('h3', '<text>', ['class' => 'class-name']);
```
Comment (markstory) [inline tests/TestCase/View/Helper/HtmlHelperTest.php]: The contents should probably be escaped by default.

File: src/View/Helper/HtmlHelper.php [hunk 2]
```diff
@@ -143,6 +143,28 @@ public function __construct(View $View, array $config = [])
...
+     *
+     * If you pass no more than two parameters, it tries to generate a html
+     *   tag with the name of the method and works as alias of `tag()`.
+     * @param string $method Method to invoke
+     * @param array $params Params for the method
+     * @return string|void
+     */
+    public function __call($method, $params)
+    {
+        if (count($params) <= 2) {
```
Comment (markstory) [inline src/View/Helper/HtmlHelper.php]: Without a whitelist of allowed tags we won't be able to easily add methods and be confident that we aren't breaking an existing application. I think a whitelist will be necessary.

**Comment by mirko-pagliai** [comment]
@markstory we had already discussed about this. See the discussion.

This method is only an alias of `tag()`. So if this method can do something wrong, it's because the `tag()` method can do something wrong.

Example. With this method, you can do something like this (**very wrong**):

    echo $this->Html->a('my link', ['href' => 'http://url']);

But I can already do the same thing:

    echo $this->Html->tag('a', 'my link', ['href' => 'http://url']);

I do not think we need a whitelist of tag. Someone might need to generate no-standard tags.

But the point is that **if we really need a whitelist of tags**, this should relate the `tag()` method.

**Comment by markstory** [comment]
> This method is only an alias of tag(). So if this method can do something wrong, it's because the tag() method can do something wrong.

The point I'm trying to get at is that having a `__call` method like this limits our ability to add methods in the *future*. Because people will be relying on the `__call` method, we will be unable to safely add any methods that match an HTML tag in the future that require a call signature different than the one `__call` supports. For example:

```php
$this->Html->video();
$this->Html->picture();
```

Will not be methods we can add except in major releases as we *could* break an existing application that relies on the `__call` semantics. This is why I would want a whitelist.

**Comment by mirko-pagliai** [comment]
Clear.

But let me say that all methods that generate a simple tag should have the same signature: `$text` and `array $options = []` args.

(from memory, at this time, only `div()` and `para()` methods have a different signature)

**Comment by thinkingmedia** [comment]
Instead of a white list. How about a method prefix?

```
   echo $this->Html->_span(....);
```

**Comment by jadb** [comment]
> Instead of a white list. How about a method prefix?

Maybe, but not an underscore (in cake those are for protected methods by convention).

**Comment by saeideng** [comment]
i'm :-1:  

instead 
```
$this->Html->aTag(....);
$this->Html->h1Tag(....);
```

> echo $this->Html->_span(....);
> 

private mode ? :lock:

**Comment by saeideng** [comment]
or create a new helper `Tag`
```
$this->Tag->a(...)
$this->Tag->h1(...)
```
//and
use Html->tag helper for created it

**Comment by jadb** [comment]
> or create a new helper Tag

as a plugin please not in core, lol.

**Comment by jadb** [comment]
Another thing to think about (since we're working nicely with IDEs auto-completion): 

With this change, *every* HTML tag would need to be documented in docblock for IDEs to parse/auto-complete `$this->Html->h3()`, etc.

Thoughts?

**Comment by saeideng** [comment]
No :)
We have many tags
How can create many tags method manualy?

**Comment by thinkingmedia** [comment]
This is more of a personal preference of how people want to work with templates, then a feature improvement for the core. The PR can't be merged because there is right way to implement this change, and there are to many possible ways people want to do it.

I say, something like this goes to a plugin.

Maybe what we need a section on the Cake website to promote plugins specifically related to templates and helpers.

**Comment by ADmad** [comment]
@thinkingmedia We already have https://github.com/FriendsOfCake/awesome-cakephp :)

**Comment by jadb** [comment]
Sorry @saeideng - not sure I understand. 

I just meant that by using a the magic `__call()`, IDEs won't be able to correctly parse magically created tags (i.e. `h3`) in views.

**Comment by mirko-pagliai** [comment]
Anyway, note that similar methods already exist in CakePHP:
http://api.cakephp.org/3.2/class-Cake.View.Helper.FormHelper.html#___call

**Pull request closed without merging**